frontend/src/metabase/dashboard/actions/data-fetching-typed.ts

```diff
@@ -210,3 +250,329 @@ export const SET_SHOW_LOADING_COMPLETE_FAVICON =
+              ? JSON.stringify(datasetQuery.parameters)
+              : undefined,
+            ignore_cache: ignoreCache,
+          },
+          queryOptions,
+        ),
+      );
+    } else if (dashboardType === "embed") {
+      result = await fetchDataOrError(
+        maybeUsePivotEndpoint(EmbedApi.dashboardCardQuery, card)(
+          {
+            token: dashcard.dashboard_id,
+            dashcardId: dashcard.id,
+            cardId: card.id,
+            parameters: JSON.stringify(
+              getParameterValuesBySlug(dashboard?.parameters, parameterValues),
+            ),
+            ignore_cache: ignoreCache,
+          },
+          queryOptions,
+        ),
+      );
+    } else if (dashboardType === "transient" || dashboardType === "inline") {
+      result = await fetchDataOrError(
+        maybeUsePivotEndpoint(MetabaseApi.dataset, card)(
+          { ...datasetQuery, ignore_cache: ignoreCache },
+          queryOptions,
+        ),
+      );
+    } else {
+      const dashcardBeforeEditing = getDashCardBeforeEditing(
+        getState(),
+        dashcard.id,
+      );
+      const hasReplacedCard =
+        dashcard.card_id != null &&
+        dashcardBeforeEditing &&
+        dashcardBeforeEditing.card_id !== dashcard.card_id;
+
+      const shouldUseCardQueryEndpoint =
+        isNewDashcard(dashcard) ||
+        (isQuestionDashCard(dashcard) &&
+          isNewAdditionalSeriesCard(card, dashcard)) ||
+        hasReplacedCard;
+
+      // new dashcards and new additional series cards aren't yet saved to the dashboard, so they need to be run using the card query endpoint
+      const endpoint = shouldUseCardQueryEndpoint
+        ? CardApi.query
+        : DashboardApi.cardQuery;
+
+      const requestBody = shouldUseCardQueryEndpoint
+        ? { cardId: card.id, ignore_cache: ignoreCache }
+        : {
+            dashboardId: dashcard.dashboard_id,
+            dashcardId: dashcard.id,
+            cardId: card.id,
+            parameters: datasetQuery.parameters,
+            ignore_cache: ignoreCache,
+            dashboard_id: dashcard.dashboard_id,
+            dashboard_load_id: dashboardLoadId,
+          };
+
+      result = await fetchDataOrError(
+        maybeUsePivotEndpoint(endpoint, card)(requestBody, queryOptions),
+      );
+    }
+
+    setFetchCardDataCancel(card.id, dashcard.id, null);
+    clearTimeout(slowCardTimer);
+
+    return {
+      dashcard_id: dashcard.id,
+      card_id: card.id,
+      result: cancelled ? null : result,
+      currentTime: performance.now(),
+    };
+  },
+);
+
+export const fetchCardData =
+  (
+    card: Card,
+    dashcard: DashboardCard,
+    options: {
+      reload?: boolean;
+      clearCache?: boolean;
+      ignoreCache?: boolean;
+      dashboardLoadId?: string;
+    } = {},
+  ) =>
+  (dispatch: Dispatch) => {
+    dispatch(fetchCardDataAction({ card, dashcard, options }));
+  };
+
+fetchCardData.fulfilled = fetchCardDataAction.fulfilled;
```

Comment:
this might be contentious. This allows us to use `createAsyncThunk` with the appropriate typings while being able to use the same call signature as `createThunkAction`. I tried to use the types available with `createThunkAction` but there were a bunch of snags that I couldn't figure out